services/nexus/nexus.service.js

```diff
@@ -1,130 +1,221 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
+const BaseJsonService = require('../base-json')
+const Joi = require('joi')
 const { isSnapshotVersion: isNexusSnapshotVersion } = require('./nexus-version')
 const { addv: versionText } = require('../../lib/text-formatters')
 const { version: versionColor } = require('../../lib/color-formatters')
+const serverSecrets = require('../../lib/server-secrets')
+
+const versionRegex = /^\d+(\.\d+)*(-.*)?$/
+
+const searchApiSchema = Joi.object({
+  data: Joi.array()
+    .items(
+      Joi.object({
+        latestRelease: Joi.string().regex(versionRegex),
+        latestSnapshot: Joi.string().regex(versionRegex),
+        version: Joi.string().regex(versionRegex),
+      })
+    )
+    .required(),
+}).required()
+
+const resolveApiSchema = Joi.object({
+  data: Joi.object({
+    baseVersion: Joi.string().regex(versionRegex),
+    version: Joi.string().regex(versionRegex),
+  }).required(),
+}).required()
+
+module.exports = class Nexus extends BaseJsonService {
+  static render({ version }) {
+    let message, color
+    if (version === 'no-artifact') {
+      message = version
+      color = 'red'
+    } else if (version !== '0') {
+      message = versionText(version)
+      color = versionColor(version)
+    } else {
+      message = 'undefined'
```

Comment:
I honestly don't know on this one (relying on the LegacyService implementation) 😄 

Based on a quick mental walkthrough of the Legacy implementation, I think this could happen under two circumstances:

1) snapshot badge is requested, but the specified group/artifact has no snapshot versions. 
2) repository badge is requested, response has no value for `baseVersion` field (which I am assuming can/does happen in real life based on [this line](https://github.com/badges/shields/blob/master/services/test-validators.js#L25) from the legacy service), nor does the response have a value for the `version` field. 

The first one feels a lot like the above scenario where throwing `NotFound` could be used. It could be covered with a schema (although it would get a little complex), but I think a pretty message conveying `no snapshot versions found` might be more helpful for the user than `invalid`. Thoughts?

The second one would be much more manageable with Joi than the first IMO, but I'm not sure what Nexus guarantees in the response. For example is `baseVersion` optional while `version` is guaranteed? Or is it that one value is guaranteed, but it could be either `version` or `baseVersion`, or neither?

Reply:
I'm familiar with what Nexus is, but haven't used it, so I'm gonna have to think about this when I'm a little fresher. If anyone else has another opinion meanwhile, please feel free to respond!

Reply:
I added some links below to the Neuxs API docs, but they didn't offer much assistance IMO. I updated the service impl. to throw `InvalidResponse` errors in each of those 2 circumstances while leaving the Joi schema a bit more flexible